vexim/sitechangesubmit.php

```diff
@@ -1,99 +1,145 @@
+
+// Detect the user action
+if (isset($_POST['allusers'])) {
+  $action = 'allusers';
+}
+elseif (isset($_POST['clear']) && $_POST['clear']!=='') {
+  $action = 'adminpass';
+} else {
+  $action = 'default';
+}
+
+// Main action switch
+switch ($action) {
+  case 'allusers':
+    // Default value
+    $success = TRUE;
+
+    // Process SpamAssassin
+    if (isset($_POST['on_spamassassin'])) {
+      if ($_POST['on_spamassassin'] != '2' && ($_POST['on_spamassassin'] == '1' || $_POST['on_spamassassin'] == '0')) {
+        $query = "UPDATE users SET on_spamassassin=:on_spamassassin WHERE domain_id=:domain_id";
+        $sth = $dbh->prepare($query);
+        $success = $sth->execute(array(
+            ':domain_id' => $_POST['allusers'],
+            ':on_spamassassin' => $_POST['on_spamassassin']));
+      }
+    }
+
+    // Process Anti-Virus
+    if (isset($_POST['on_avscan']) && $success) {
+      if ($_POST['on_avscan'] != '2' && ($_POST['on_avscan'] == '1' || $_POST['on_avscan'] == '0')) {
+        $query = "UPDATE users SET on_avscan=:on_avscan WHERE domain_id=:domain_id";
+        $sth = $dbh->prepare($query);
+        $success = $sth->execute(array(
+            ':domain_id' => $_POST['allusers'],
+            ':on_avscan' => $_POST['on_avscan']));
+      }
+    }
+    break;
+
+  case 'adminpass':
     if (validate_password($_POST['clear'], $_POST['vclear'])) {
       $query = "UPDATE users SET crypt=:crypt WHERE localpart=:localpart AND domain_id=:domain_id";
       $sth = $dbh->prepare($query);
-      $success = $sth->execute(array(':crypt'=>crypt_password($_POST['clear']),
-            ':localpart'=>$_POST['localpart'], ':domain_id'=>$_POST['domain_id']));
-      if ($success) {
-	header ("Location: site.php?updated={$_POST['domain']}");
-	die;
-      } else {
-	header ("Location: site.php?failupdated={$_POST['domain']}");
-	die;
-      }
+      $success = $sth->execute(array(
+          ':crypt' => crypt_password($_POST['clear']),
+          ':localpart' => $_POST['localpart'],
+          ':domain_id' => $_POST['domain_id']));
     } else {
-      header ("Location: site.php?badpass={$_POST['domain']}");
+      header("Location: site.php?badpass={$_POST['domain']}");
       die;
     }
-  }
+    break;
+
+  default:
+    if (isset($_POST['avscan'])) {
+      $_POST['avscan'] = 1;
+    } else {
+      $_POST['avscan'] = 0;
+      $query = "UPDATE users SET on_avscan=:on_avscan WHERE domain_id=:domain_id";
+      $sth = $dbh->prepare($query);
+      $sth->execute(array(
+            ':domain_id' => $_POST['domain_id'],
+            ':on_avscan' => '0'));
+    }
+
+    if (isset($_POST['spamassassin'])) {
+      $_POST['spamassassin'] = 1;
+    } else {
+      $_POST['spamassassin'] = 0;
+      $query = "UPDATE users SET on_spamassassin=:on_spamassassin WHERE domain_id=:domain_id";
+      $sth = $dbh->prepare($query);
+      $sth->execute(array(
+        ':domain_id' => $_POST['domain_id'],
+        ':on_spamassassin' => '0'));
+    }
```

Comment:
The problem I had was: If you disable spamassassin or avscan for the domain and you don't disable it for all users, exim continues filtering spam/virus-mails. The spam router (https://github.com/vexim/vexim2/blob/master/docs/configure#L518-L525) doesn't verify if spamassassin is enabled for the domain itself.
So in this case, the additional feature to turn on/of SA/AV for all users only makes sense if you enable SA/AV later and want to activate it for all existing users.

Alternative: in the exim-router, we check `domains.on_spamassassin=1`. I would prefer this option, because if you turn of SA for a domain by accident, you can easily turn it on again without the users loosing their settings.